### PR TITLE
Plumb shape detection API to the GPU Process

### PR DESCRIPTION
**Pull request by litherum**
#### f2195aa52bec637d3a35273da68217f83e845350
<pre>
Plumb shape detection API to the GPU Process
<a href="https://bugs.webkit.org/show_bug.cgi?id=255473">https://bugs.webkit.org/show_bug.cgi?id=255473</a>
rdar://108070184

Reviewed by Mike Wyrzykowski.

This patch is really just plumbing, though there are a few pieces to it.

1. The shape detection API is already behind a runtime flag, which guards the JavaScript entry points.
       So, the code is getting compiled everywhere. However, eventually something somewhere has to
       actually implement the API itself (for Cocoa ports, that&apos;s the Vision framework), and not all
       ports have this support. So, there also needs to be a build-time flag,
       HAVE(SHAPE_DETECTION_API_IMPLEMENTATION), which _just_ guards the calls to the final
       implementation. This is just to make the code compile on all ports.
2. WebCore doesn&apos;t know anything about IPC, so the object creation methods need to be plumbed up
       through ChromeClient, so WebKitLegacy or WebKit2 can choose whether to back the objects by
       WebKit::ShapeDetection::RemoteFooProxy objects, or by WebCore::ShapeDetection::FooImpl objects.
       Because these object creation methods are free functions in JavaScript, they need to have the
       [CallWith=CurrentScriptExecutionContext] annotation, to get the context they need to talk to
       the Page and ChromeClient.
3. WebCore::ShapeDetection::FooImpl objects need to be exported from WebCore, so WebKit and
       WebKitLegacy can create them
4. Create RemoteFooProxy and RemoteFoo classes. RemoteFooProxy classes live in the web process, and
       forward all their method calls to IPC send(). RemoteFoo classes live in the GPU process, and own
       an instance of the corresponding WebCore::ShapeDetection::FooImpl object. There are messages
       defined in RemoteFoo.messages.in files.
5. Handle object lifetime in the GPU process. When an object is created in the web process, that
       corresponds to a RemoteRenderingBackend::CreateFoo() message. The RemoteRenderingBackend owns a
       WebKit::ShapeDetection::ObjectHeap, which retains all the objects in the GPU process that get
       created. The objects that get created start listening on the preexisting StreamConnection of the
       RemoteRenderingBackend - I didn&apos;t want to give each shape detection object its own totally new
       StreamConnection (like how WebGPU does) because these shape detection objects are so small and
       only have 1 or 2 methods on them. Object destructors in the web process send a
       RemoteRenderingBackend::ReleaseFoo() message.
6. Add serialization information for the relevant structs to WebCoreArgumentCoders.serialization.in

The last thing that&apos;s worth mentioning is I haven&apos;t hooked up the ImageBitmapSource parameters to the
detect() functions, because those sources can be in many different forms, and the logic to handle all
the forms is quite large, and outside the scope of this simple plumbing patch. My next patch will
implement these parameters.

No tests because there is no behavior change - I haven&apos;t actually hooked up the Vision framework to
actually implement any of the calls yet.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/Modules/ShapeDetection/BarcodeDetector.cpp:
(WebCore::BarcodeDetector::create):
(WebCore::BarcodeDetector::getSupportedFormats):
* Source/WebCore/Modules/ShapeDetection/BarcodeDetector.h:
* Source/WebCore/Modules/ShapeDetection/BarcodeDetector.idl:
* Source/WebCore/Modules/ShapeDetection/FaceDetector.cpp:
(WebCore::FaceDetector::create):
* Source/WebCore/Modules/ShapeDetection/FaceDetector.h:
* Source/WebCore/Modules/ShapeDetection/FaceDetector.idl:
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/BarcodeDetectorImplementation.h:
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/FaceDetectorImplementation.h:
* Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/TextDetectorImplementation.h:
* Source/WebCore/Modules/ShapeDetection/Interfaces/BarcodeDetectorOptionsInterface.h:
* Source/WebCore/Modules/ShapeDetection/Interfaces/BarcodeFormatInterface.h:
* Source/WebCore/Modules/ShapeDetection/Interfaces/DetectedBarcodeInterface.h:
* Source/WebCore/Modules/ShapeDetection/Interfaces/DetectedFaceInterface.h:
* Source/WebCore/Modules/ShapeDetection/Interfaces/DetectedTextInterface.h:
* Source/WebCore/Modules/ShapeDetection/Interfaces/LandmarkInterface.h:
* Source/WebCore/Modules/ShapeDetection/Interfaces/LandmarkTypeInterface.h:
* Source/WebCore/Modules/ShapeDetection/TextDetector.cpp:
(WebCore::TextDetector::create):
* Source/WebCore/Modules/ShapeDetection/TextDetector.h:
* Source/WebCore/Modules/ShapeDetection/TextDetector.idl:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::createBarcodeDetector const):
(WebCore::Chrome::getBarcodeDetectorSupportedFormats const):
(WebCore::Chrome::createFaceDetector const):
(WebCore::Chrome::createTextDetector const):
* Source/WebCore/page/Chrome.h:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::createBarcodeDetector const):
(WebCore::ChromeClient::getBarcodeDetectorSupportedFormats const):
(WebCore::ChromeClient::createFaceDetector const):
(WebCore::ChromeClient::createTextDetector const):
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/ShapeDetection/RemoteBarcodeDetector.cpp: Copied from Source/WebCore/Modules/ShapeDetection/TextDetector.h.
(WebKit::RemoteBarcodeDetector::RemoteBarcodeDetector):
(WebKit::RemoteBarcodeDetector::detect):
* Source/WebKit/GPUProcess/ShapeDetection/RemoteBarcodeDetector.h: Added.
(WebKit::RemoteBarcodeDetector::create):
(WebKit::RemoteBarcodeDetector::backing):
* Source/WebKit/GPUProcess/ShapeDetection/RemoteBarcodeDetector.messages.in: Added.
* Source/WebKit/GPUProcess/ShapeDetection/RemoteFaceDetector.cpp: Copied from Source/WebCore/Modules/ShapeDetection/TextDetector.h.
(WebKit::RemoteFaceDetector::RemoteFaceDetector):
(WebKit::RemoteFaceDetector::detect):
* Source/WebKit/GPUProcess/ShapeDetection/RemoteFaceDetector.h: Added.
(WebKit::RemoteFaceDetector::create):
(WebKit::RemoteFaceDetector::backing):
* Source/WebKit/GPUProcess/ShapeDetection/RemoteFaceDetector.messages.in: Added.
* Source/WebKit/GPUProcess/ShapeDetection/RemoteTextDetector.cpp: Copied from Source/WebCore/Modules/ShapeDetection/TextDetector.h.
(WebKit::RemoteTextDetector::RemoteTextDetector):
(WebKit::RemoteTextDetector::detect):
* Source/WebKit/GPUProcess/ShapeDetection/RemoteTextDetector.h: Added.
(WebKit::RemoteTextDetector::create):
(WebKit::RemoteTextDetector::backing):
* Source/WebKit/GPUProcess/ShapeDetection/RemoteTextDetector.messages.in: Added.
* Source/WebKit/GPUProcess/ShapeDetection/ShapeDetectionObjectHeap.cpp: Copied from Source/WebCore/Modules/ShapeDetection/TextDetector.cpp.
(WebKit::ShapeDetection::ObjectHeap::addObject):
(WebKit::ShapeDetection::ObjectHeap::removeObject):
(WebKit::ShapeDetection::ObjectHeap::clear):
* Source/WebKit/GPUProcess/ShapeDetection/ShapeDetectionObjectHeap.h: Copied from Source/WebCore/Modules/ShapeDetection/Implementation/Cocoa/BarcodeDetectorImplementation.h.
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::RemoteRenderingBackend):
(WebKit::RemoteRenderingBackend::createRemoteBarcodeDetector):
(WebKit::RemoteRenderingBackend::releaseRemoteBarcodeDetector):
(WebKit::RemoteRenderingBackend::getRemoteBarcodeDetectorSupportedFormats):
(WebKit::RemoteRenderingBackend::createRemoteFaceDetector):
(WebKit::RemoteRenderingBackend::releaseRemoteFaceDetector):
(WebKit::RemoteRenderingBackend::createRemoteTextDetector):
(WebKit::RemoteRenderingBackend::releaseRemoteTextDetector):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/webrtc/RemoteAudioMediaStreamTrackRendererInternalUnitManager.cpp:
(WebKit::RemoteAudioMediaStreamTrackRendererInternalUnitManager::Unit::captureUnitHasStopped):
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
(headers_for_type):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/ShapeDetectionIdentifier.h: Copied from Source/WebCore/Modules/ShapeDetection/Interfaces/BarcodeDetectorOptionsInterface.h.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteBarcodeDetectorProxy.cpp: Added.
(WebKit::ShapeDetection::RemoteBarcodeDetectorProxy::create):
(WebKit::ShapeDetection::RemoteBarcodeDetectorProxy::RemoteBarcodeDetectorProxy):
(WebKit::ShapeDetection::RemoteBarcodeDetectorProxy::~RemoteBarcodeDetectorProxy):
(WebKit::ShapeDetection::RemoteBarcodeDetectorProxy::getSupportedFormats):
(WebKit::ShapeDetection::RemoteBarcodeDetectorProxy::detect):
* Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteBarcodeDetectorProxy.h: Added.
(WebKit::ShapeDetection::RemoteBarcodeDetectorProxy::backing const):
* Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteFaceDetectorProxy.cpp: Added.
(WebKit::ShapeDetection::RemoteFaceDetectorProxy::create):
(WebKit::ShapeDetection::RemoteFaceDetectorProxy::RemoteFaceDetectorProxy):
(WebKit::ShapeDetection::RemoteFaceDetectorProxy::~RemoteFaceDetectorProxy):
(WebKit::ShapeDetection::RemoteFaceDetectorProxy::detect):
* Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteFaceDetectorProxy.h: Added.
(WebKit::ShapeDetection::RemoteFaceDetectorProxy::backing const):
* Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteTextDetectorProxy.cpp: Added.
(WebKit::ShapeDetection::RemoteTextDetectorProxy::create):
(WebKit::ShapeDetection::RemoteTextDetectorProxy::RemoteTextDetectorProxy):
(WebKit::ShapeDetection::RemoteTextDetectorProxy::~RemoteTextDetectorProxy):
(WebKit::ShapeDetection::RemoteTextDetectorProxy::detect):
* Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteTextDetectorProxy.h: Added.
(WebKit::ShapeDetection::RemoteTextDetectorProxy::backing const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.h:
* Source/WebKit/WebProcess/GPU/webrtc/SampleBufferDisplayLayer.h:
* Source/WebKit/WebProcess/Inspector/WebInspectorUIExtensionController.cpp:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createBarcodeDetector const):
(WebKit::WebChromeClient::getBarcodeDetectorSupportedFormats const):
(WebKit::WebChromeClient::createFaceDetector const):
(WebKit::WebChromeClient::createTextDetector const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.cpp:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::createBarcodeDetector const):
(WebChromeClient::getBarcodeDetectorSupportedFormats const):
(WebChromeClient::createFaceDetector const):
(WebChromeClient::createTextDetector const):

Canonical link: <a href="https://commits.webkit.org/263094@main">https://commits.webkit.org/263094@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/068e3f23c3a64b6bc748ec4bf794fac8d551499f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3603 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3644 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3792 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5029 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/3867 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3577 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3744 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3692 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/3126 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3647 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/3892 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3249 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4851 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/3634 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/1398 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3222 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/3279 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/2942 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3199 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3281 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4615 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/3378 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3656 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/2958 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/3675 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3191 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3222 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/939 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/874 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3222 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/3767 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3471 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1050 "Passed tests") |

### CODE REVIEW

File: Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteTextDetectorProxy.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    RemoteTextDetectorProxy(RemoteTextDetectorProxy&&) = delete;
+    RemoteTextDetectorProxy& operator=(const RemoteTextDetectorProxy&) = delete;
+    RemoteTextDetectorProxy& operator=(RemoteTextDetectorProxy&&) = delete;
+
+    ShapeDetectionIdentifier backing() const { return m_backing; }
+
+    void detect(CompletionHandler<void(Vector<WebCore::ShapeDetection::DetectedText>&&)>&&) final;
+
+    ShapeDetectionIdentifier m_backing;
+    RemoteRenderingBackendProxy& m_remoteRenderingBackendProxy;
```
Comment (mwyrzykowski) [inline Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteTextDetectorProxy.h]: Should we use `WeakPtr<>` like in `RemoteDisplayListRecorderProxy` or `RemoteImageBufferProxy`?

There was previously a similar crash where RemoteGPUProxy was destructed before other Proxy classes, leading to use of a reference after its destructor was called

File: Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteFaceDetectorProxy.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    RemoteFaceDetectorProxy(RemoteFaceDetectorProxy&&) = delete;
+    RemoteFaceDetectorProxy& operator=(const RemoteFaceDetectorProxy&) = delete;
+    RemoteFaceDetectorProxy& operator=(RemoteFaceDetectorProxy&&) = delete;
+
+    ShapeDetectionIdentifier backing() const { return m_backing; }
+
+    void detect(CompletionHandler<void(Vector<WebCore::ShapeDetection::DetectedFace>&&)>&&) final;
+
+    ShapeDetectionIdentifier m_backing;
+    RemoteRenderingBackendProxy& m_remoteRenderingBackendProxy;
```
Comment (mwyrzykowski) [inline Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteFaceDetectorProxy.h]: same comment as https://github.com/WebKit/WebKit/pull/12765/files#r1167344105

File: Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteBarcodeDetectorProxy.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    RemoteBarcodeDetectorProxy(RemoteBarcodeDetectorProxy&&) = delete;
+    RemoteBarcodeDetectorProxy& operator=(const RemoteBarcodeDetectorProxy&) = delete;
+    RemoteBarcodeDetectorProxy& operator=(RemoteBarcodeDetectorProxy&&) = delete;
+
+    ShapeDetectionIdentifier backing() const { return m_backing; }
+
+    void detect(CompletionHandler<void(Vector<WebCore::ShapeDetection::DetectedBarcode>&&)>&&) final;
+
+    ShapeDetectionIdentifier m_backing;
+    RemoteRenderingBackendProxy& m_remoteRenderingBackendProxy;
```
Comment (mwyrzykowski) [inline Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteBarcodeDetectorProxy.h]: similar to https://github.com/WebKit/WebKit/pull/12765/files#r1167344105

**Comment by mwyrzykowski** [comment]
Other than the one comment and Linux build failures, this change seems fine to me.

**Comment by litherum** [comment]
https://github.com/litherum/WebKit/tree/ShapeDetectionImageBitmapSource is where the patch that hooks up the ImageBitmapSources is.

**Comment by litherum** [comment]
And https://github.com/litherum/WebKit/tree/ShapeDetectionImplementation is where the patch that actually implements the API on top of the Vision framework

File: Source/WebKit/GPUProcess/ShapeDetection/ShapeDetectionObjectHeap.h [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+private:
+    ObjectHeap();
+
+    using Object = std::variant<
+        std::monostate,
+        Ref<RemoteBarcodeDetector>,
+        Ref<RemoteFaceDetector>,
+        Ref<RemoteTextDetector>
+    >;
+    HashMap<ShapeDetectionIdentifier, Object> m_objects;
```
Comment (kkinnunen-apple) [inline Source/WebKit/GPUProcess/ShapeDetection/ShapeDetectionObjectHeap.h]: I don't think it makes sense to add complexity in order to have one hash map. The solution is inferior in every single aspect to having one map per type.
The type automatically divides the object key space naturally.

File: Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteBarcodeDetectorProxy.cpp [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+#if ENABLE(GPU_PROCESS)
+
+#include "MessageSenderInlines.h"
+#include "RemoteBarcodeDetectorMessages.h"
+#include "RemoteRenderingBackendProxy.h"
+#include "WebProcess.h"
+
+namespace WebKit::ShapeDetection {
+
+Ref<RemoteBarcodeDetectorProxy> RemoteBarcodeDetectorProxy::create(RemoteRenderingBackendProxy& remoteRenderingBackendProxy, ShapeDetectionIdentifier identifier, const WebCore::ShapeDetection::BarcodeDetectorOptions& barcodeDetectorOptions)
```
Comment (kkinnunen-apple) [inline Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteBarcodeDetectorProxy.cpp]: If the classes don't need the RemoteRenderingBackendProxy, you could only store the the stream connection.

(this has implications wrt how these objects would handle gpu crashes and crash recovery)

File: Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteBarcodeDetectorProxy.h [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    RemoteBarcodeDetectorProxy(RemoteBarcodeDetectorProxy&&) = delete;
+    RemoteBarcodeDetectorProxy& operator=(const RemoteBarcodeDetectorProxy&) = delete;
+    RemoteBarcodeDetectorProxy& operator=(RemoteBarcodeDetectorProxy&&) = delete;
+
+    ShapeDetectionIdentifier backing() const { return m_backing; }
+
+    void detect(CompletionHandler<void(Vector<WebCore::ShapeDetection::DetectedBarcode>&&)>&&) final;
+
+    ShapeDetectionIdentifier m_backing;
+    IPC::StreamClientConnection& m_streamClientConnection;
```
Comment (litherum) [inline Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteBarcodeDetectorProxy.h]: This should be `Ref<>`

**Review by mwyrzykowski — APPROVED** [review]

**Pull request merged**